File: ShortestWordDistance3.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
# Using Boolean flag if the both the words are same then update index2 to index1


class Solution:
    def shortestWordDistance(self, wordsDict: List[str], word1: str,
                             word2: str) -> int:
        if not wordsDict:
            return None

        index1 = -1
        index2 = -1
        minDistance = inf
        if word1 == word2:
            flag = True
        else:
            flag = False
        for i in range(len(wordsDict)):
            if word1 == wordsDict[i]:
                if flag:
                    index2 = index1
                index1 = i
            elif word2 == wordsDict[i]:
                index2 = i
            if index1 != -1 and index2 != -1:
                minDistance = min(minDistance, abs(index2 - index1))
        return minDistance

File: assignments/src/ShortestWordDistance_3.java
// Approach: Two pointers and Min calc everytime. Return the final Min value.
// Note: as both words can be same, make a check - if (p1 == i) -> p1 = p2.
// Time: O(n)
// Space: O(1)

class ShortestWordDistance_3 {
    public int shortestWordDistance(String[] wordsDict, String word1, String word2) {
        int p1 = -1, p2 = -1, min = Integer.MAX_VALUE;

        for (int i = 0; i<wordsDict.length; i++) {
            String currWord = wordsDict[i];

            if (currWord.equals(word1)) {
                p1 = i;
            }
            if (currWord.equals(word2)) {
                if (p1 == i) {          // will happen only if word1 = word2
                    p1 = p2;
                }
                p2 = i;
            }
            if (p1 != -1 && p2 != -1) {
                min = Math.min(min, Math.abs(p1-p2));
            }
        }
        return min;
    }
}